### PR TITLE
feat: add create user statement

### PR DESCRIPTION
**Pull request by dl239**
add new statements as below:
- create user
- drop user
- alter user

### CODE REVIEW

File: zetasql/parser/bison_parser.y [hunk 1]
```diff
@@ -2366,6 +2417,28 @@ opt_as_sql_function_body_or_string:
       }
     ;
 
+identifier_or_string_list:
+    identifier_or_string
```
Comment (aceforeverd) [inline zetasql/parser/bison_parser.y]: I think simple **identifier** is enough, no need for strings.

File: zetasql/parser/bison_parser.y [hunk 2]
```diff
@@ -2197,6 +2222,32 @@ function_parameters:
...
+        $$ = MAKE_NODE(ASTUserList, @$, {$1});
+      }
+    | user_list "," user_info
+      {
+        $$ = WithEndLocation(WithExtraChildren($1, {$3}), @$);
+      }
+    ;
+
+user_info:
+    identifier_or_string opt_password
```
Comment (aceforeverd) [inline zetasql/parser/bison_parser.y]: we can make options followed by **identifier** more extensible. `IDENTIFIED BY` is good but I guess **options_list** better:
```yacc
CREATE USER name OPTIONS ( password = ..., others = ... )
```
And `PASSWORD` do not have to be keyword.

File: zetasql/parser/testdata/alter_user.test
Comment (aceforeverd) [inline zetasql/parser/testdata/alter_user.test:1]: or 
```yacc
ALTER USER name [ IF EXISTS ] SET OPTION options_list
```
?

File: zetasql/parser/parse_tree_manual.h [hunk 3]
```diff
@@ -4648,6 +4648,117 @@ class ASTCreateTableStmtBase : public ASTCreateStatement {
...
+  void InitFields() final {
+    FieldLoader fl(this);
+    fl.AddRequired(&user_name_list_);
+  }
+
+  const ASTIdentifierOrStringList* user_name_list_ = nullptr;
+  bool is_if_exists_ = false;
+};
+
+class ASTAlterUserStatement final : public ASTStatement {
```
Comment (aceforeverd) [inline zetasql/parser/parse_tree_manual.h]: inherit `ASTAlterStatementBase`

File: zetasql/parser/parse_tree_manual.h [hunk 4]
```diff
@@ -4648,6 +4648,117 @@ class ASTCreateTableStmtBase : public ASTCreateStatement {
...
+  void InitFields() final {
+    FieldLoader fl(this);
+    fl.AddRequired(&user_list_);
+  }
+
+  bool is_if_exists_ = false;
+  const ASTUserList* user_list_ = nullptr;
+};
+
+class ASTCreateUserStatement final : public ASTStatement {
```
Comment (aceforeverd) [inline zetasql/parser/parse_tree_manual.h]: inheriting `ASTCreateStatement`

File: zetasql/parser/parse_tree_manual.h [hunk 5]
```diff
@@ -4648,6 +4648,117 @@ class ASTCreateTableStmtBase : public ASTCreateStatement {
   const ASTPathExpression* like_table_name_ = nullptr;       // May be NULL.
 };
 
+class ASTDropUserStatement final : public ASTStatement {
```
Comment (aceforeverd) [inline zetasql/parser/parse_tree_manual.h]: inherit `ASTDdlStatement`

File: zetasql/parser/bison_parser.y (hunk 2 above)
Comment (dl239) [reply zetasql/parser/bison_parser.y]: it's better to use the existing syntax like mysql/tidb.

File: zetasql/parser/parse_tree_manual.h (hunk 3 above)
Comment (dl239) [reply zetasql/parser/parse_tree_manual.h]: the `path_` member of `ASTAlterStatementBase` is path_expression

File: zetasql/parser/bison_parser.y (hunk 1 above)
Comment (dl239) [reply zetasql/parser/bison_parser.y]: we may support IP control as 'user'@'IP' later. so users may create by strings as 
`create user 'user1';`

File: zetasql/parser/bison_parser.y (hunk 2 above)
Comment (aceforeverd) [reply zetasql/parser/bison_parser.y]: it's not extensible, CREATE USER if not SQL standard, syntax is left to implementation.

File: zetasql/parser/bison_parser.y (hunk 1 above)
Comment (aceforeverd) [reply zetasql/parser/bison_parser.y]: I think all other thing can left to `options_list` like discussed below

File: zetasql/parser/parse_tree_manual.h (hunk 3 above)
Comment (aceforeverd) [reply zetasql/parser/parse_tree_manual.h]: u can create standalone filed in inherited class

Comment (dl239) [reply zetasql/parser/parse_tree_manual.h]: so what's the meaning of inherit?  there are some required method in base class we need?

File: zetasql/parser/bison_parser.y (hunk 2 above)
Comment (dl239) [reply zetasql/parser/bison_parser.y]: we cannot add option at the end of  the list? if there are other concerns, we can limit creating one user at most

Comment (aceforeverd) [reply zetasql/parser/bison_parser.y]: pls check 'create table' 's usage of options. Our primary target SQl dialect is, more likely SparkSQL, there is no need to follow a non-standard implementation from MySQL. 

'Create User' should introduce more features (at least I think), with options_list u don't have to update zetasql each time.

File: zetasql/parser/parse_tree_manual.h (hunk 3 above)
Comment (aceforeverd) [reply zetasql/parser/parse_tree_manual.h]: it's **base** class, u should follow the existing framework as possible as u can

File: zetasql/parser/bison_parser.y (hunk 2 above)
Comment (dl239) [reply zetasql/parser/bison_parser.y]: the user experience and habits are the primary consideration. i do not think adding non standard syntax at first is a good decision

Comment (aceforeverd) [reply zetasql/parser/bison_parser.y]: MySQL is not standard. postgresql has bit different syntax https://www.postgresql.org/docs/current/sql-createrole.html. I doubt we should consider burden from MySQL. 

Instead consider more about extensible and unified SQL interface

File: zetasql/parser/parse_tree_manual.h (hunk 3 above)
Comment (dl239) [reply zetasql/parser/parse_tree_manual.h]: `ASTAlterAllRowAccessPoliciesStatement` has not inherited the `ASTAlterStatementBase`.    a class inherit a base class means the derived class have the same behavior or attribute with the base class. the `path_` and `action_list_` in the `ASTAlterStatementBase` are the required filed that `ASTAlterUserStatement` is not needed

File: zetasql/parser/bison_parser.y (hunk 2 above)
Comment (dl239) [reply zetasql/parser/bison_parser.y]: ok

File: zetasql/parser/parse_tree_manual.h (hunk 3 above)
Comment (aceforeverd) [reply zetasql/parser/parse_tree_manual.h]: @dl239 So I think the syntax should better: `alter user name [if not exists] set options optiions_list` , just discussed in https://github.com/4paradigm/zetasql/pull/52#discussion_r1420051755. Then we can use the `ASTSetOptionsAction`

Bacause hybridse engine have a `AltertActionBase` too, and if we can unified here, hybridse's code became cleaner too. 
https://github.com/4paradigm/OpenMLDB/blob/25a24dcb12aa68dc961c2587729b2221ab349a7d/hybridse/src/planv2/ast_node_converter.cc#L2341-L2370

Altert action should have a ACTION name, like `ADD | SET | DROP | RENAME`

Comment (dl239) [reply zetasql/parser/parse_tree_manual.h]: ok

File: zetasql/parser/parse_tree_manual.h [hunk 6]
```diff
@@ -4648,6 +4648,47 @@ class ASTCreateTableStmtBase : public ASTCreateStatement {
...
+  void InitFields() final {
+    FieldLoader fl(this);
+    fl.AddRequired(&name_);
+  }
+
+  const ASTPathExpression* name_;
+  bool is_if_exists_ = false;
+};
+
+class ASTCreateUserStatement final : public ASTCreateTableStmtBase {
```
Comment (aceforeverd) [inline zetasql/parser/parse_tree_manual.h]: `ASTCreateTableStmtBase` -> `ASTCreateStatement`

**Review by aceforeverd — APPROVED** [review]

**Pull request closed without merging**